sdks/java/io/fileschematransform/src/main/java/org/apache/beam/sdk/io/fileschematransform/FileWriteSchemaTransformConfiguration.java

```diff
@@ -0,0 +1,215 @@
+
+    /** The number of output shards produced; a value of 1 disables sharding. */
+    public abstract Builder setNumShards(Integer value);
+
+    /** Uses the given {@link org.apache.beam.sdk.io.ShardNameTemplate} for naming output files. */
+    public abstract Builder setShardNameTemplate(String value);
+
+    /** Configures the filename suffix for written files. */
+    public abstract Builder setFilenameSuffix(String value);
+
+    /** Configures extra details related to writing CSV formatted files. */
+    public abstract Builder setCsvConfiguration(CsvConfiguration value);
+
+    /** Configures extra details related to writing Parquet formatted files. */
+    public abstract Builder setParquetConfiguration(ParquetConfiguration value);
+
+    /** Configures extra details related to writing XML formatted files. */
+    public abstract Builder setXmlConfiguration(XmlConfiguration value);
+
+    public abstract FileWriteSchemaTransformConfiguration build();
+  }
+
+  /** Configures extra details related to writing CSV formatted files. */
+  @DefaultSchema(AutoValueSchema.class)
+  @AutoValue
+  public abstract static class CsvConfiguration {
+
+    /** Content written to the top of every sharded file prior to the header. */
+    @Nullable
+    public abstract String getPreamble();
+
+    /**
+     * The format of the written CSV file. See <a
+     * href="https://commons.apache.org/proper/commons-csv/apidocs/org/apache/commons/csv/CSVFormat.html">org.apache.commons.csv.CSVFormat</a>
+     * for allowed values, stringified in lowercase. Defaults to <a
+     * href="https://commons.apache.org/proper/commons-csv/apidocs/org/apache/commons/csv/CSVFormat.html#DEFAULT">CSVFormat.DEFAULT</a>
+     */
+    @Nullable
+    public abstract String getCsvFormat();
+
+    @AutoValue.Builder
+    public abstract static class Builder {
+
+      /** Content written to the top of every sharded file prior to the header. */
+      public abstract Builder setPreamble(String value);
+
+      /**
+       * The format of the written CSV file. See <a
+       * href="https://commons.apache.org/proper/commons-csv/apidocs/org/apache/commons/csv/CSVFormat.html">org.apache.commons.csv.CSVFormat</a>
+       * for allowed values, stringified in lowercase. Defaults to <a
+       * href="https://commons.apache.org/proper/commons-csv/apidocs/org/apache/commons/csv/CSVFormat.html#DEFAULT">CSVFormat.DEFAULT</a>
+       */
+      public abstract Builder setCsvFormat(String value);
+
+      public abstract CsvConfiguration build();
+    }
+  }
+
+  /** Configures extra details related to writing Parquet formatted files. */
+  @DefaultSchema(AutoValueSchema.class)
+  @AutoValue
+  public abstract static class ParquetConfiguration {
+
+    /**
+     * Specifies compression codec. See org.apache.parquet.hadoop.metadata.CompressionCodecName for
+     * allowed names, stringified in lowercase format.
+     */
+    public abstract String getCompressionCodecName();
```

Comment:
shouldn't this be set from the higher-level `compression` attribute in the `FileWriteSTConf..`?

Reply:
The FileWriteSchemaTransformProvider is essentially a factory for FileIO/TextIO given a format.  The design goals of the configuration are essentially to unify the various parameters of the different format based FileIOs such as XmlIO, ParquetIO, etc.  To keep it simple, it adheres as close as possible to a Sink's parameters.  Given the primitive type constraint of SchemaTransformProvider's configurations, we need to accept a string and convert to the enumeration type.  For `getCompressionCodecName` it is https://www.javadoc.io/doc/org.apache.parquet/parquet-hadoop/1.8.1/org/apache/parquet/hadoop/metadata/CompressionCodecName.html

Reply:
makes sense thanks